kategory-optics/src/main/kotlin/kategory/optics/Traversal.kt

```diff
@@ -0,0 +1,337 @@
+package kategory.optics
+
+import kategory.Applicative
+import kategory.Const
+import kategory.Either
+import kategory.HK
+import kategory.Id
+import kategory.IntMonoid
+import kategory.ListKW
+import kategory.Monoid
+import kategory.Option
+import kategory.applicative
+import kategory.identity
+import kategory.left
+import kategory.map
+import kategory.monoid
+import kategory.none
+import kategory.right
+import kategory.some
+import kategory.traverse
+import kategory.value
+
+/**
+ * [Traversal] is a type alias for [PTraversal] which fixes the type arguments
+ * and restricts the [PTraversal] to monomorphic updates.
+ */
+typealias Traversal<S, A> = PTraversal<S, S, A, A>
+
+/**
+ * A [Traversal] is an optic that allows to see into a structure with 0 to N targets.
+ *
+ * [Traversal] is a generalisation of [kategory.Traverse] and can be seen as a representation of modifyF.
+ * all methods are written in terms of modifyF
+ *
+ * @param S the source of a [PTraversal]
+ * @param T the modified source of a [PTraversal]
+ * @param A the target of a [PTraversal]
+ * @param B the modified target of a [PTraversal]
+ */
+abstract class PTraversal<S, T, A, B> {
```

Comment:
![](https://i.imgur.com/V7tqRN6.png)

Reply:
Currently all optics are build with abstract classes since the polymorphic PR. I could make them interfaces. I can actually not remember why I went for `abstract class` instead of interface.

Reply:
We've been using interfaces for typeclasses and such, so if there's no real reason for them to be abstract classes then yes, please do the refactor.

Reply:
I will! Then I can also make the modifyF method which is currently abstract internal and only expose the reified function.  👍